tests/components/songpal/test_config_flow.py

```diff
@@ -0,0 +1,249 @@
+        interfaceVersion="interface version",
+        serverName="server name",
+    )
+
+
+def _create_mocked_device(throw_exception=False):
+    mocked_device = MagicMock()
+    type(mocked_device).get_supported_methods = MagicMock(
+        side_effect=_get_supported_methods(throw_exception)
+    )
+    type(mocked_device).get_interface_information = MagicMock(
+        side_effect=_get_interface_information
+    )
+    return mocked_device
+
+
+def _patch_config_flow_device(mocked_device):
+    return patch(
+        "homeassistant.components.songpal.config_flow.Device",
+        return_value=mocked_device,
+    )
+
+
+def _flow_next(hass, flow_id):
+    return next(
+        flow
+        for flow in hass.config_entries.flow.async_progress()
+        if flow["flow_id"] == flow_id
+    )
+
+
+async def test_flow_ssdp(hass):
+    """Test working ssdp flow."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": SOURCE_SSDP}, data=SSDP_DATA,
+    )
+    assert result["type"] == "form"
+    assert result["step_id"] == "init"
+    assert result["description_placeholders"] == {
+        CONF_NAME: FRIENDLY_NAME,
+        CONF_HOST: HOST,
+    }
+    flow = _flow_next(hass, result["flow_id"])
+    assert flow["context"]["unique_id"] == UDN
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"], user_input={}
+    )
+    assert result["type"] == RESULT_TYPE_CREATE_ENTRY
+    assert result["title"] == FRIENDLY_NAME
+    assert result["data"] == CONF_DATA
+
+
+async def test_flow_user(hass):
+    """Test working user initialized flow."""
+    mocked_device = _create_mocked_device()
+
+    with _patch_config_flow_device(mocked_device):
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_USER},
+        )
+        assert result["type"] == RESULT_TYPE_FORM
+        assert result["step_id"] == "user"
+        assert result["errors"] is None
+        _flow_next(hass, result["flow_id"])
+
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"], user_input={CONF_ENDPOINT: ENDPOINT},
+        )
+        assert result["type"] == RESULT_TYPE_CREATE_ENTRY
+        assert result["title"] == MODEL
+        assert result["data"] == {
+            CONF_NAME: MODEL,
+            CONF_ENDPOINT: ENDPOINT,
+        }
+
+    mocked_device.get_supported_methods.assert_called_once()
+    mocked_device.get_interface_information.assert_called_once()
+
+
+async def test_flow_import(hass):
+    """Test working import flow."""
+    mocked_device = _create_mocked_device()
+
+    with _patch_config_flow_device(mocked_device):
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=CONF_DATA
+        )
+        assert result["type"] == RESULT_TYPE_CREATE_ENTRY
+        assert result["title"] == FRIENDLY_NAME
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5462445)